### ISSUE TITLE
Different predictions from XGBoost Model trained with QuantileDMatrix on same dataset

### ISSUE THREAD

**Issue by mustaq95** [dmlc/xgboost#9788]
@trivialfis @abhishek0093 Can you please look into my issue once. I have added the [link](https://colab.research.google.com/drive/14T7dZyFr5VSw-dY0qd30Hl9iGiVxmkHW#scrollTo=vePZtWXIs5Mt) to reproduce the issue.
The issue is that i'm observing different quantile predictions values when model is trained with different column orders (sorting alphabetically) QuantileDMatrix for a given alpha list.
This is not ensuring model consistency even when trained on same dataset. Please help me in fixing this.

**Comment by trivialfis** [comment]
I can take a closer look tomorrow to check if there's an actual error. But I wouldn't be surprised if there's none. During tree split enumeration, the ordering of the input can sometimes affect the output model due to floating point accuracy.

**Comment by mustaq95** [comment]
If it's so, then can you help me in reproducing same output after sorting as well. In my case the model feature importance is also getting changed for both models.

**Comment by mustaq95** [comment]
@trivialfis Have you checked upon this issue?

**Comment by trivialfis** [comment]
looking.

**Comment by trivialfis** [comment]
I think this is a floating point induced error. I looked into the internal splitting process, XGBoost can generate near identical loss gain between splitting on "retailer_id" and "category_id", which is not surprising since they come from the same probability distribution. When there's a tie-breaking decision for splitting a node, the ordering of the feature comes into play.

**Comment by mustaq95** [comment]
@trivialfis is there any way i can generate the same output independent of the order of the columns with quantileerror loss function. Please can you have a deeper look into this or suggest an example which show same output with sorted columns order as well.

**Comment by trivialfis** [comment]
> is there any way i can generate the same output independent of the order of the columns with quantileerror loss function

This sort of numerical issue is usually considered normal and is not tied to any loss function. I don't think there's a general solution to it. The easiest way to workaround it is through feature engineering, like always sorting the columns of the input before training a model so that you can keep the model input deterministic.

**Comment by mustaq95** [comment]
this sort of changes the feature importance of the model as well (gives weight to earlier columns in my case) and the model becomes indeterministic.

**Comment by trivialfis** [comment]
Yes, it's one of the issues. So far, when we define "deterministic model", we mean: given the exact same input data with the same software and hardware environment, we can produce an exact same model for multiple runs.

Anything other than that is not guaranteed. We can probably define an extra RV to uniformly pick a random split when there are tie breaking candidates. But that seems to be over engineering at the moment as there are many things that computer not being consistent with real number maths, for instance, change of row other can also change final models.

**Comment by mustaq95** [comment]
so that means there is no way to get the same model results after sorting the columns.

**Comment by trivialfis** [comment]
Not without pulling some tricks like differentiating features by sacle and mean, or just sort all datasets. But in general, no, there's no general solution that can work with every dataset without these tricks.

Out of curiosity, if we were to internally sort the columns before training, would it count as deterministic output from your perspective?

**Comment by mustaq95** [comment]
yes, that will definitely work. This will prevent in creating different model versions on the same data with different columns order and consistent feature importance.

**Comment by trivialfis** [comment]
Great, then you can resolve it by creating a wrapper of the Booster class that feeds data into the booster after sorting!

**Comment by trivialfis** [comment]
Feel free to reopen if there are further questions.
